File: cracking-the-coding-interview/interview-questions/linked_lists/singly_linked_list.py
class Node:
	def __init__(self, data=None):
		self.data = data
		self.next = None

class LinkedList:
	def __init__(self):
		self.head = None
		self.size = 0
		
	def insert(self, data):
		node = Node(data)
		if self.head == None:
			self.head = node
			self.size += 1
			return
		curr = self.head
		while curr.next:
			curr = curr.next
		curr.next = node
		curr = node
		self.size += 1

	def delete(self, data):
		if self.head == None: return
		curr = self.head
		prev = self.head
		while curr.next:
			if curr.data == data:
				prev.next = curr.next
				self.size -= 1
				return
			prev = curr
			curr = curr.next

	def search(self, data):
		curr = self.head
		cnt = 0
		while curr.next and curr.data != data:
			cnt += 1
			curr = curr.next
		if curr.data == data: return cnt
		return -1

	def get(self, index):
		curr = self.head
		cnt = 0
		while curr.next and cnt != index:
			cnt += 1
			curr = curr.next
		if cnt == index: return curr.data
		return None


	def get_size(self):
		print(self.size)

	def to_string(self):
		curr = self.head
		while curr:
			print(curr.data, end=' -> ')
			curr = curr.next
		print('None')

if __name__ == "__main__":
	linked_list = LinkedList()
	linked_list.insert(5)
	linked_list.insert(6)
	linked_list.insert(4)
	linked_list.insert(7)
	linked_list.insert(8)
	linked_list.insert(19)
	linked_list.to_string()
	linked_list.get_size()
	linked_list.delete(6)
	linked_list.to_string()
	print(linked_list.search(8))
	print(linked_list.get(3))
	print(linked_list.search(19))

File: data-struc-impl/linkedlist/SinglyLinkedList.java
import java.util.List;
import java.util.ArrayList;

public class SinglyLinkedList<T> {
	public static class Node<T> {
		private T data;
		private Node<T> next;

		public Node() {
			this.data = null;
			this.next = null;
		}

		public Node(T data) {
			this.data = data;
		}

		public Node(T data, Node<T> next) {
			this.data = data;
			this.next = next;
		}
	}

	private Node<T> head;
	private int size;

	public SinglyLinkedList() {
		this.head = null;
	}

	public void insert(T data) {
		Node<T> insert = new Node<>(data);
		if (this.head == null) {
			this.head = insert;
			size++;
			return;
		}
		Node<T> curr = this.head;
		while (curr.next != null) {
			curr = curr.next;
		}
		curr.next = insert;
		size++;
	}

	public void delete(T data) {
		if (this.head == null)
			return;
		Node<T> curr = this.head;
		Node<T> prev = this.head;
		while (curr.next != null) {
			if (curr.data == data) {
				prev.next = curr.next;
				size--;
				return;
			}
			prev = curr;
			curr = curr.next;
		}
	}

	public int search(T data) {
		Node<T> curr = this.head;
		int cnt = 0;
		while (curr.next != null && curr.data != data) {
			cnt++;
			curr = curr.next;
		}
		if (curr.data == data)
			return cnt;
		return -1;
	}

	public T get(int index) {
		Node<T> curr = this.head;
		int cnt = 0;
		while (curr.next != null && cnt != index) {
			cnt++;
			curr = curr.next;
		}
		if (cnt == index)
			return curr.data;
		return null;
	}

	public int size() {
		return this.size;
	}

	public String toString() {
		List<T> list = new ArrayList<>();
		Node<T> curr = this.head;
		while (curr != null) {
			list.add(curr.data);
			curr = curr.next;
		}
		return list.toString();
	}

	public static void main(String[] args) {
		SinglyLinkedList<Integer> ll = new SinglyLinkedList<>();
		System.out.println(ll.toString());
		ll.insert(4);
		System.out.println(ll.toString());
		ll.insert(2);
		ll.insert(18);
		ll.insert(43);
		ll.insert(993);
		ll.insert(34);
		System.out.println(ll.toString());
		System.out.println(ll.search(18));
		System.out.println(ll.get(2));
		System.out.println(ll.size());
		System.out.println(ll.get(4));
		System.out.println(ll.search(993));
	}
}